File: sdk/python/tests/test_error_taxonomy.py
from __future__ import annotations

import json
import sys
import unittest
from pathlib import Path

PACKAGE_ROOT = Path(__file__).resolve().parents[1]
REPO_ROOT = PACKAGE_ROOT.parents[1]
sys.path.insert(0, str(PACKAGE_ROOT / "src"))

from glyphastore.client import (  # noqa: E402
    Client,
    MutationOutcome,
    _enrich,
    _retryability_for,
)


def load_cases() -> list[dict]:
    path = PACKAGE_ROOT / "tests" / "fixtures" / "error_taxonomy_v1.json"
    if not path.is_file():
        path = REPO_ROOT / "tests" / "fixtures" / "error_taxonomy_v1.json"
    data = json.loads(path.read_text(encoding="utf-8"))
    return list(data["cases"])


class ErrorTaxonomyTests(unittest.TestCase):
    def test_wire_status_category_retryability_matrix(self) -> None:
        for case in load_cases():
            with self.subTest(case["id"]):
                status = case["wire_status"]
                error = Client._status_error(status)
                self.assertEqual(error.category, case["category"])
                self.assertEqual(error.wire_status, case["wire_status"])
                self.assertEqual(error.retryability, case["read_retryability"])

                outcome = MutationOutcome(case["mutation_outcome"])
                enriched = _enrich(
                    Client._status_error(status),
                    bytes_sent=1,
                    mutation_outcome=outcome,
                    wire_status=case["wire_status"],
                )
                self.assertEqual(enriched.retryability, case["mutation_retryability"])
                self.assertEqual(
                    _retryability_for(
                        case["category"],
                        True,
                        case["mutation_outcome"] == "indeterminate",
                    ),
                    case["mutation_retryability"],
                )
                self.assertEqual(
                    case["unhealthy"],
                    case["wire_status"] in (6, 7),
                )


if __name__ == "__main__":
    unittest.main()

File: sdk/ruby/test/test_error_taxonomy.rb
# frozen_string_literal: true

require "json"
require "pathname"
require "minitest/autorun"
require "glypha_store"

class TestErrorTaxonomy < Minitest::Test
  FIXTURES = Pathname.new(__dir__).join("fixtures")
  REPO_FIXTURES = Pathname.new(__dir__).join("../../../tests/fixtures")

  def fixture(name)
    path = FIXTURES.join(name)
    path = REPO_FIXTURES.join(name) unless path.file?
    JSON.parse(path.read)
  end

  def test_wire_status_category_retryability_matrix
    fixture("error_taxonomy_v1.json").fetch("cases").each do |case_item|
      error = GlyphaStore::Error.from_status(case_item.fetch("wire_status"))
      assert_equal case_item.fetch("category"), error.category, case_item["id"]
      assert_equal case_item.fetch("wire_status"), error.wire_status, case_item["id"]
      assert_equal case_item.fetch("read_retryability"), error.retryability, case_item["id"]

      enriched = GlyphaStore::Error.from_status(case_item.fetch("wire_status")).enrich(
        bytes_sent: 1,
        mutation_outcome: case_item.fetch("mutation_outcome")
      )
      assert_equal case_item.fetch("mutation_retryability"), enriched.retryability, case_item["id"]

      want_unhealthy =
        case_item.fetch("wire_status") == GlyphaStore::Protocol::Status::WRONG_OWNER ||
        case_item.fetch("wire_status") == GlyphaStore::Protocol::Status::NOT_BOUND
      assert_equal want_unhealthy, case_item.fetch("unhealthy"), case_item["id"]
    end
  end

  def test_indeterminate_enrich_ignores_zero_bytes_sent
    # Receive-after-send paths historically omitted bytes_sent; transport+0 must not
    # advertise same_request when mutation_outcome is already indeterminate.
    error = GlyphaStore::Error.transport("socket closed")
                              .enrich(mutation_outcome: GlyphaStore::MutationOutcome::INDETERMINATE)
    assert_equal 0, error.bytes_sent
    assert_equal GlyphaStore::Retryability::RECONCILE_FIRST, error.retryability
    assert_equal GlyphaStore::MutationOutcome::INDETERMINATE, error.mutation_outcome
  end
end
